src/azul/terra.py

```diff
@@ -159,11 +159,15 @@ class TerraClient:
     @cached_property
     def credentials(self) -> Credentials:
         with aws.service_account_credentials() as file_name:
-            return Credentials.from_service_account_file(file_name)
+            credentials = Credentials.from_service_account_file(file_name)
+        credentials = credentials.with_scopes(self.credential_scopes)
+        credentials.refresh(Request())  # Obtain access token
+        return credentials
 
     oauth_scopes = [
         'email',
         'openid',
+        'https://www.googleapis.com/auth/bigquery.readonly',
```

Comment:
I tried deleting this line and running the repository proxy tests and it still passed. Maybe I'm running the wrong tests, but I'm thinking again this line should be deleted.

Reply:
The unit tests load canned tables into TinyQuery and do not make any actual requests to BigQuery. The quickest way to verify that the client has sufficient authorizations is to run the `make sam` target:

```
$ git diff
diff --git a/src/azul/terra.py b/src/azul/terra.py
index 33315a3a..01df75d5 100644
--- a/src/azul/terra.py
+++ b/src/azul/terra.py
@@ -255,7 +255,6 @@ class TDRClient(SAMClient):
         return [
             *super().credential_scopes,
             'https://www.googleapis.com/auth/devstorage.read_only',
-            'https://www.googleapis.com/auth/bigquery.readonly'
         ]
 
     @cache
$ make sam
python /home/nadove/PycharmProjects/azul-sc/scripts/register_sam.py
2021-05-06 20:20:10,064 DEBUG   MainThread: _oauthed_request('POST', 'https://sam.dsde-dev.broadinstitute.org/register/user/v1', fields=None, headers=None, body='')
2021-05-06 20:20:10,085 INFO    MainThread: Found credentials in environment variables.
2021-05-06 20:20:11,132 DEBUG   MainThread: _oauthed_request(…) -> b'{"causes":[],"message":"user CreateWorkbenchUser(2620357612338d02b515c,116556158276818438280,azul-ucsc-4-nadove@platform-hca-dev.iam.gserviceaccount.com,None) already exists","source":"sam","stackTrace":[],"statusCode":409}'
2021-05-06 20:20:11,132 INFO    MainThread: Google service account previously registered with SAM.
2021-05-06 20:20:11,134 DEBUG   MainThread: _oauthed_request('GET', 'https://jade.datarepo-dev.broadinstitute.org/api/repository/v1/snapshots', fields={'filter': 'lungmap_dev_20210412__20210414', 'limit': 2}, headers=None, body=None)
2021-05-06 20:20:11,590 DEBUG   MainThread: _oauthed_request(…) -> b'{"total":1,"items":[{"id":"215793b0-b991-4b73-9584-1103d4072f39","name":"lungmap_dev_20210412__20210414","description":"Create snapshot lungmap_dev_20210412__20210414","createdDate":"2021-04-14T19:45:16.972386Z","profileId":"390e7a85-d47f-4531-b612-165f...'
2021-05-06 20:20:11,591 DEBUG   MainThread: _oauthed_request('GET', 'https://jade.datarepo-dev.broadinstitute.org/api/repository/v1/snapshots/215793b0-b991-4b73-9584-1103d4072f39', fields=None, headers=None, body=None)
2021-05-06 20:20:12,002 DEBUG   MainThread: _oauthed_request(…) -> b'{"id":"215793b0-b991-4b73-9584-1103d4072f39","name":"lungmap_dev_20210412__20210414","description":"Create snapshot lungmap_dev_20210412__20210414","createdDate":"2021-04-14T19:45:16.972386Z","source":[{"dataset":{"id":"10330a3a-d0e8-41c7-bc1a-2a5cb02c8...'
2021-05-06 20:20:12,003 DEBUG   MainThread: _oauthed_request('GET', 'https://jade.datarepo-dev.broadinstitute.org/api/repository/v1/snapshots', fields={'filter': 'lungmap_dev_20210412__20210414', 'limit': 2}, headers=None, body=None)
2021-05-06 20:20:12,163 DEBUG   MainThread: _oauthed_request(…) -> b'{"total":1,"items":[{"id":"215793b0-b991-4b73-9584-1103d4072f39","name":"lungmap_dev_20210412__20210414","description":"Create snapshot lungmap_dev_20210412__20210414","createdDate":"2021-04-14T19:45:16.972386Z","profileId":"390e7a85-d47f-4531-b612-165f...'
2021-05-06 20:20:12,165 DEBUG   MainThread: _oauthed_request('GET', 'https://jade.datarepo-dev.broadinstitute.org/api/repository/v1/snapshots/215793b0-b991-4b73-9584-1103d4072f39', fields=None, headers=None, body=None)
2021-05-06 20:20:12,617 DEBUG   MainThread: _oauthed_request(…) -> b'{"id":"215793b0-b991-4b73-9584-1103d4072f39","name":"lungmap_dev_20210412__20210414","description":"Create snapshot lungmap_dev_20210412__20210414","createdDate":"2021-04-14T19:45:16.972386Z","source":[{"dataset":{"id":"10330a3a-d0e8-41c7-bc1a-2a5cb02c8...'
2021-05-06 20:20:12,618 INFO    MainThread: TDR client is authorized for API access to tdr:broad-jade-dev-data:snapshot/lungmap_dev_20210412__20210414:.
Traceback (most recent call last):
  File "/home/nadove/PycharmProjects/azul-sc/src/azul/terra.py", line 319, in check_bigquery_access
    tables = list(bigquery.list_tables(source.bq_name, max_results=1))
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/google/api_core/page_iterator.py", line 212, in _items_iter
    for page in self._page_iter(increment=False):
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/google/api_core/page_iterator.py", line 243, in _page_iter
    page = self._next_page()
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/google/api_core/page_iterator.py", line 369, in _next_page
    response = self._get_next_page_response()
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/google/api_core/page_iterator.py", line 418, in _get_next_page_response
    return self.api_request(
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/google/cloud/bigquery/client.py", line 556, in _call_api
    return call()
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/google/api_core/retry.py", line 281, in retry_wrapped_func
    return retry_target(
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/google/api_core/retry.py", line 184, in retry_target
    return target()
  File "/home/nadove/PycharmProjects/azul-sc/.venv/lib/python3.8/site-packages/google/cloud/_http.py", line 483, in api_request
    raise exceptions.from_http_response(response)
google.api_core.exceptions.Forbidden: 403 GET https://bigquery.googleapis.com/bigquery/v2/projects/broad-jade-dev-data/datasets/lungmap_dev_20210412__20210414/tables?maxResults=1&prettyPrint=false: Request had insufficient authentication scopes.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/nadove/PycharmProjects/azul-sc/scripts/register_sam.py", line 43, in <module>
    main()
  File "/home/nadove/PycharmProjects/azul-sc/scripts/register_sam.py", line 39, in main
    tdr.check_bigquery_access(source)
  File "/home/nadove/PycharmProjects/azul-sc/src/azul/terra.py", line 330, in check_bigquery_access
    raise self._insufficient_access(resource)
azul.RequirementError: The service account (SA) 'azul-ucsc-4-nadove@platform-hca-dev.iam.gserviceaccount.com' is not authorized to access BigQuery dataset 'lungmap_dev_20210412__20210414' in Google Cloud project 'broad-jade-dev-data' or that resource does not exist. Make sure that it exists, that the SA is registered with SAM and has been granted read access to the resource.
Makefile:91: recipe for target 'sam' failed
make: *** [sam] Error 1
```